src/state_machine.zig

```diff
@@ -403,17 +403,51 @@ pub fn StateMachineType(comptime Storage: type) type {
 
             var chain: ?usize = null;
```

Comment:
For consistency, let's rely on the existing chain rollback mechanism. That way, only the last event in the chain has the specific error, and all others are `linked_event_failed`.

This diff is based on the original state machine:

```diff
--- a/src/state_machine.zig
+++ b/src/state_machine.zig
@@ -409,11 +409,13 @@ pub fn StateMachineType(comptime Storage: type) type {
                     chain = index;
                     assert(chain_broken == false);
                 }
-                const result = if (chain_broken) .linked_event_failed else switch (operation) {
-                    .create_accounts => self.create_account(event),
-                    .create_transfers => self.create_transfer(event),
-                    else => unreachable,
-                };
+                const result = if (chain_broken) .linked_event_failed
+                    else if (event.flags.linked and index == events.len - 1) .malformed_linked_chain
+                    else switch (operation) {
+                        .create_accounts => self.create_account(event),
+                        .create_transfers => self.create_transfer(event),
+                        else => unreachable,
+                    };
                 log.debug("{s} {}/{}: {}: {}", .{
                     @tagName(operation),
                     index + 1,
@@ -448,10 +450,6 @@ pub fn StateMachineType(comptime Storage: type) type {
                     chain_broken = false;
                 }
             }
-            // TODO client.zig: Validate that batch chains are always well-formed and closed.
-            // This is programming error and we should raise an exception for this in the client ASAP.
-            assert(chain == null);
-            assert(chain_broken == false);

             return @sizeOf(Result(operation)) * count;
         }
```

Reply:
@sentientwaffle, by "rely on the existing chain rollback mechanism"

Did you mean letting the whole chain execute normally and then rollbacking at the end when we detected the `chain != null`?
I had started doing it that way, but I assumed this rollback could be expensive and better if avoided.

Reply:
> Did you mean letting the whole chain execute normally and then rollbacking at the end when we detected the `chain != null`?

Yes, rolling back to the start of the (invalid) chain.

>  I had started doing it that way, but I assumed this rollback could be expensive and better if avoided.

This case is rare — it is only hit when the client is used wrong (i.e. a bug in the application code). So the performance of the rollback isn't an issue.

Reply:
There is a separate performance cost that comes in, which is the additional branch on every event, to test flags and index. I don't know if we can ride on another condition to get rid of this also? For 8k events in a batch, with such a hot loop, this kind of branching is significant. The less conditions in the data plane here the better.